surveys/question_types.py

```diff
@@ -0,0 +1,137 @@
+
+
+__all__ = ['QuestionType', 'CharQuestion', 'IntQuestion', 'ChoiceQuestion', 'get_question_type']
+
+
+class QuestionType:
+    """
+    QuestionTypes are the bridge between the database Question types and the front end visualisation
+
+    Database related parameters:
+    type_slug: slugified version for database type storage
+    type_verbose: readable name for the question type
+    options: - tbd -
+
+    Form related paramters
+    formfield_class: The field class for the form
+    widget: the widget used for displaying. Defaults to formfield_class default
+    """
+    formfield_class = None
+    type_slug = None
+    type_verbose = None
+    options = {}
+    widget = None
+
+    def __init__(self):
+        # Todo: Move to metaclass
+        assert self.formfield_class is not None
+        assert self.type_slug is not None
+        assert self.type_verbose is not None
+
+    def get_form_field(self, question, answer=None):
+        """
+        Returns the form field instance
+        :param question: The question for the form field
+        :param answer: The answer instance
+        :return:
+        """
+        return self.formfield_class(
+            initial=self.get_field_answer(answer),
+            label=question.label,
+            widget=self.widget,
+            required=question.required,
+            **self.get_extra_form_field_kwargs(question, answer=answer)
+        )
+
+    def get_extra_form_field_kwargs(self, question, answer=None):
+        """
+        Additional form field kwargs used for form field initialisation.
+        Use this for setting field specific options (such as choices, min_value etc)
+        """
+        return {
+            'validators': self.get_field_validators(question),
+        }
+
+    def get_field_validators(self, question):
+        """ Returns additional validators """
+        return []
+
+    def get_field_answer(self, answer):
+        """ Converts db stored answer to answer used for field communication """
+        if answer is None:
+            return ''
+        return answer.value
+
+    def answer_to_db(self, question, value):
+        """ Converts the field answer to database ready answer """
+        return value
+
+
+class CharQuestion(QuestionType):
+    """ A simple open question """
+    formfield_class = fields.CharField
+    type_slug = 'CHAR'
+    type_verbose = 'Open answer'
+
+
+class IntQuestion(QuestionType):
+    """ An integer question """
+    formfield_class = fields.IntegerField
+    type_slug = 'INT'
+    type_verbose = 'Integer answer'
+
+    def get_extra_form_field_kwargs(self, question, answer=None):
+        kwargs = super(IntQuestion, self).get_extra_form_field_kwargs(question, answer=answer)
+        if question.option_1:
+            kwargs['min_value'] = int(question.option_1)
+        if question.option_2:
+            kwargs['max_value'] = int(question.option_2)
+        return kwargs
+
+
+class ChoiceQuestion(QuestionType):
```

Comment:
A multi-choice question (with checkboxes) could be a worthy addition.

Reply:
I am intending to alter type registration a bit more, maybe in a too elaborate way, where each type will be represented in a QuestionType table. This way we can introduce new forms in the future if neccessary